eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Binary/CostModelsSpec.hs

```diff
@@ -0,0 +1,156 @@
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeApplications #-}
+
+module Test.Cardano.Ledger.Alonzo.Binary.CostModelsSpec (spec) where
+
+import Cardano.Ledger.Alonzo.Core
+import Cardano.Ledger.Alonzo.Scripts
+import Cardano.Ledger.BaseTypes
+import Cardano.Ledger.Binary (ToCBOR (..))
+import Cardano.Ledger.Binary.Decoding
+import Cardano.Ledger.Binary.Encoding
+import Cardano.Ledger.Plutus.Language
+import qualified Data.Map as Map
+import Data.Text (Text)
+import Data.Typeable
+import Data.Word (Word8)
+import Lens.Micro
+import Test.Cardano.Ledger.Alonzo.Arbitrary
+import Test.Cardano.Ledger.Common
+
+spec ::
+  forall era.
+  AlonzoEraPParams era =>
+  Proxy era ->
+  Spec
+spec pxy = do
+  describe "CostModels CBOR deserialization" $ do
+    validCostModelProp @era pxy
+    invalidCostModelProp @era Proxy
+    unknownCostModelProp @era Proxy
+
+validCostModelProp ::
+  forall era.
+  AlonzoEraPParams era =>
+  Proxy era ->
+  Spec
+validCostModelProp pxy = do
+  prop "valid CostModels deserialize correctly, both independently and within PParamsUpdate" $
+    \(lang :: Language) -> do
+      forAllShow (genValidCostModelsEnc lang) (showEnc pxy) $
+        \validCmEnc -> do
+          encodeAndCheckDecoded @era validCmEnc $
+            \cmDecoded ppuDecoded -> do
+              expectRight cmDecoded >>= (`shouldSatisfy` validCm)
+              ppuRes <- expectRight ppuDecoded
+              ppuRes `shouldSatisfy` \ppu -> (validCm <$> ppu ^. ppuCostModelsL) == SJust True
+  where
+    genValidCostModelsEnc lang = genCostModelsEncForLanguage lang (costModelParamsCount lang)
+    validCm (CostModels valid errors unknown) =
+      (not . null) valid && null errors && null unknown
+
+invalidCostModelProp ::
+  forall era.
+  AlonzoEraPParams era =>
+  Proxy era ->
+  Spec
+invalidCostModelProp pxy = do
+  prop "invalid CostModels fail within PParamsUpdate" $
+    \(lang :: Language) -> do
+      forAllShow (genInvalidCostModelsEnc lang) (showEnc pxy) $
+        \invalidCmEnc -> do
+          encodeAndCheckDecoded @era invalidCmEnc $
+            \cmDecoded ppuDecoded -> do
+              -- pre-Conway we are failing when deserializing invalid costmodels
+              if eraProtVerHigh @era < natVersion @9
+                then expectDeserialiseFailure cmDecoded (Just "CostModels")
+                else do
+                  -- post-Conway, we are collecting CostModels deserialization errors
+                  cmRes <- expectRight cmDecoded
+                  cmRes `shouldSatisfy` invalidCm
+
+              -- in no era are we deserializing invalid costmodels within PParamsUpdate
+              expectDeserialiseFailure ppuDecoded Nothing
+  where
+    genInvalidCostModelsEnc lang = do
+      let validCount = costModelParamsCount lang
+      count <- choose (0, validCount - 1)
+      genCostModelsEncForLanguage lang count
+
+    invalidCm (CostModels valid errors _) =
+      null valid && (not . null) errors
```

Comment:
Ditto redundant composition. Also should the unknown be empty here as well?
```suggestion
    invalidCm (CostModels valid errors unknown) =
      null valid && not (null errors) && null unknown
```

Reply:
This is actually an interesting case. On one hand there is no possibility for an invalid CM for a known language to all of a sudden become valid, so we could discard it. On the other keeping it in memory not only allows us to debug a problem if there is one, but also prevents us from violating roundtrip property.
So, I think it is best to preserve the current behavior, which means you can ignore my suggestion here.

Thank you for bringing this to my attention, since I haven't actually thought about this case before.